### ISSUE TITLE
dart2js: Fails to compile constructor with assert in initializer list

### ISSUE THREAD

**Issue by eernstg** [dart-lang/sdk#31321]
All subtests other than `none` are passing for the following runs:
```
tools/test.py --host-checked -mrelease -cdart2js -rd8 --dart2js-with-kernel --dart2js_options="--library-root=out/ReleaseX64/dart-sdk/" language_2/assertion_initializer_const_error2
```
The test is a multi-test where cases other than `none` are all expected to have a compile-time error, but this is a `!$checked` configuration and the test is designed to incur these compile-time errors because an `assert` is failing.

So, by design, this test has
```
assertion_initializer_const_error2_test/none: Pass
assertion_initializer_const_error2_test/*: MissingCompileTimeError # By design.
```
for all `!$checked` configurations (for all tools), because there should be a compile-time error if and only if this is a `$checked` run.

However, the above runs generate a different error:
```
Error: Constructor 'C.cc06' is not a valid constant constructor.
  const C.cc06(x, y) : x = x
        ^
Error: Compilation failed.
```
So I've put the following into the status file:
```
assertion_initializer_const_error2_test/*: Pass # Issue #31321
```
to indicate that it is passing, and it should not do that.

Please ensure that dart2js stops reporting `not a valid constant constructor` and then change that status entry to be as follows (in `language_2/language_2_dart2js.status`):
```
[ $compiler == dart2js && $dart2js_with_kernel && !$checked ]
assertion_initializer_const_error2_test/none: Pass
assertion_initializer_const_error2_test/*: MissingCompileTimeError # By design
...
```

**Comment by rakudrama** [comment]
This is blocked on: https://github.com/dart-lang/sdk/issues/30914
We can't handle assert initializers during constant evaluation until we have a resolution on what Kernel structure to expect for assert in initializer position. @kmillikin

**Comment by anders-sandholm** [comment]
#30914 has been closed - "unblocking"

**Comment by anders-sandholm** [comment]
@sigmundch - would be great to find an owner for this on dart2js.

**Comment by sigmundch** [comment]
Fixed by Johnni at f2b8cafb444b9c1b8d32e8ed7c5fc4e1ea714704
